hazelcast/src/main/java/com/hazelcast/sql/impl/FieldsUtil.java

```diff
@@ -156,18 +159,36 @@ private static boolean skipMethod(@Nonnull Class<?> clazz, @Nonnull Method metho
         for (String name : clazz.getFieldNames()) {
             FieldType portableType = clazz.getFieldType(name);
 
-            QueryDataType type = resolvePortableType(portableType);
+            QueryDataType type = resolveType(portableType);
 
             fields.putIfAbsent(name, type);
         }
 
         return fields;
     }
 
+    /**
+     * Resolve the list of fields from a schema {@link com.hazelcast.internal.serialization.impl.compact.Schema},
+     * along with their {@link QueryDataType}.
+     */
+    @Nonnull
+    public static SortedMap<String, QueryDataType> resolveCompact(@Nonnull Schema schema) {
+        SortedMap<String, QueryDataType> fields = new TreeMap<>();
+
+        // Add regular fields.
+        for (String name : schema.getFieldNames()) {
+            FieldType compactType = schema.getField(name).getType();
+            QueryDataType type = resolveType(compactType);
+            fields.putIfAbsent(name, type);
```

Comment:
Minor: Is it possible to have duplicate field names in the schema?

Reply:
You can not pass same field name twice in the `CREATE MAPPING`, it is already checked somewhere else. 
`SqlCreateMapping.validate`